mysticeti-core/src/consensus/universal_committer.rs

```diff
@@ -26,20 +27,29 @@ impl UniversalCommitter {
     /// Try to commit part of the dag. This function is idempotent and returns a list of
     /// ordered decided leaders.
     #[tracing::instrument(skip_all, fields(last_decided = %last_decided))]
-    pub fn try_commit(&self, last_decided: BlockReference) -> Vec<LeaderStatus> {
+    pub fn try_commit(&self, last_decided: AuthorityRound) -> Vec<LeaderStatus> {
         let highest_known_round = self.block_store.highest_round();
         // let last_decided_round = max(last_decided.round(), 1); // Skip genesis.
         let last_decided_round = last_decided.round();
         let last_decided_round_authority = (last_decided.round(), last_decided.authority);
 
         // Try to decide as many leaders as possible, starting with the highest round.
         let mut leaders = VecDeque::new();
-        for round in (last_decided_round..=highest_known_round).rev() {
+        'outer: for round in (last_decided_round..=highest_known_round.saturating_sub(2)).rev() {
             for committer in self.committers.iter().rev() {
                 // Skip committers that don't have a leader for this round.
                 let Some(leader) = committer.elect_leader(round) else {
                     continue;
                 };
+
+                // now that we reached the last committed leader we can stop the commit rule
+                if (round, leader) == last_decided_round_authority {
+                    tracing::debug!(
+                        "Leader of round {round} -> {leader} - reached last committed, now exit"
+                    );
+                    break 'outer;
+                }
```

Comment:
mmh.. should we invert those checks? otherwise, are we sure we are not going to miss the `last_decided_round_authority` simply because its `round` is not a leader round?

Reply:
Actually, why not move this check before the loop line 39 (`for committer in self.committers.iter().rev()`)?

Reply:
> mmh.. should we invert those checks? otherwise, are we sure we are not going to miss the last_decided_round_authority simply because its round is not a leader round?

since we always start the loop from the `last_decided_round` then then `round` eventually it will be a leader round. 

> Actually, why not move this check before the loop line 39 (for committer in self.committers.iter().rev())?

we need the leader information though don't we? Also in case we have multiple leaders per round we need to make sure that we stop to the right leader

Reply:
ha yes you're right, I forgot again that that the order of last committed leaders is defined by both (round, leader) rather than only rounds